File: julia/11_min_max_median.jl
# min_max_median(lst): a function that takes a simple list of numbers lst
# as a parameter and returns a list with the min, max, and the median of lst.
# Can you devise an algorithm that has an expected linear running time?

# Quickselect helper: partition arr[lo:hi] around the element at pivot_index
function partition!(arr, lo, hi, pivot_index)
    pivot = arr[pivot_index]
    # Move pivot to end
    arr[pivot_index], arr[hi] = arr[hi], arr[pivot_index]
    store = lo
    for i in lo:hi-1
        if arr[i] < pivot
            arr[store], arr[i] = arr[i], arr[store]
            store += 1
        end
    end
    # Move pivot to its final place
    arr[store], arr[hi] = arr[hi], arr[store]
    return store
end

# Quickselect: finds the k-th smallest element (1-based) in arr[lo:hi]
function quickselect!(arr, lo, hi, k)
    if lo == hi
        return arr[lo]
    end
    # pick a random pivot between lo and hi
    p = rand(lo:hi)
    p = partition!(arr, lo, hi, p)
    if k == p
        return arr[k]
    elseif k < p
        return quickselect!(arr, lo, p-1, k)
    else
        return quickselect!(arr, p+1, hi, k)
    end
end

"""
    min_max_median(lst)

Return `[min, max, median]` of the numeric list `lst`.  
Runs in O(n) expected time using Quickselect for the median.
"""
function min_max_median(lst)
    n = length(lst)
    n == 0 && error("List must be non-empty")

    # 1) Find min and max in one pass
    mn = lst[1]
    mx = lst[1]
    for x in lst
        if x < mn; mn = x; end
        if x > mx; mx = x; end
    end

    # 2) Prepare a copy for median selection
    arr = [x for x in lst]   # simple element‐by‐element copy
    k   = div(n + 1, 2)      # middle index, 1-based

    # 3) Quickselect to find the median
    med = quickselect!(arr, 1, n, k)

    return [mn, mx, med]
end


# Example usage:
nums = [7, 2, 9, 4, 3, 8, 1]
println(min_max_median(nums))  # [1, 9, 4]

nums1 = [7, 2, 9, 4, 3]
println(min_max_median(nums1))   # [2, 9, 4]

nums2 = [10, 1, 5, 3]
println(min_max_median(nums2))   # [1.0, 10.0, 4.0] (medians are 3 & 5, average = 4.0)

File: python/11_min_max_median.py
# min_max_median(lst): a function that takes a simple list of numbers lst as a
# parameter and returns a list with the min, max, and the median of lst. Can
# you devise an algorithm that has an expected linear running time?

# Leetcode 215. Kth Largest Element in an Array
# https://leetcode.com/problems/kth-largest-element-in-an-array/

import random

# Helper function
def quickselect(a, k):
    """
    Return the k-th smallest element (0-based) from list a.
    Uses in-place Quickselect, so we work on a copy.
    Expected O(n) time.
    """
    arr = list(a)  # work on a copy
    def select(lo, hi, k):
        if lo == hi:
            return arr[lo]
        # pick a random pivot index
        p = random.randint(lo, hi)
        # partition around arr[p]
        pivot = arr[p]
        # move pivot to end
        arr[p], arr[hi] = arr[hi], arr[p]
        store = lo
        for i in range(lo, hi):
            if arr[i] < pivot:
                arr[store], arr[i] = arr[i], arr[store]
                store += 1
        # move pivot to its final place
        arr[store], arr[hi] = arr[hi], arr[store]
        # store is pivot index
        if k == store:
            return arr[store]
        elif k < store:
            return select(lo, store - 1, k)
        else:
            return select(store + 1, hi, k)
    return select(0, len(arr) - 1, k)

def min_max_median(lst):
    """
    Returns [min, max, median] of lst.
    Median is the middle element if len is odd,
    or the average of the two middle elements if even.
    Expected O(n) time.
    """
    n = len(lst)
    if n == 0:
        raise ValueError("List must be non-empty")

    # 1) Find min and max in one pass
    mn = mx = lst[0]
    for x in lst[1:]:
        if x < mn: mn = x
        if x > mx: mx = x

    # 2) Find median via Quickselect
    if n % 2 == 1:
        # odd: single middle element
        med = quickselect(lst, n // 2)
    else:
        # even: average of lower and upper middles
        m1 = quickselect(lst, n//2 - 1)
        m2 = quickselect(lst, n//2)
        med = (m1 + m2) / 2

    return [mn, mx, med]

# --- Example Usage ---
lst1 = [7, 2, 9, 4, 3, 8, 1]
print(min_max_median(lst1))  # [1, 9, 4]

lst2 = [5, 2, 1, 4]
print(min_max_median(lst2))  # [1, 5, 3.0]

lst3 = [10]
print(min_max_median(lst3))  # [10, 10, 10]
